src/hooks/useFetch.js

```diff
@@ -4,103 +4,86 @@
-  const [data, setData] = React.useState(null);
-  const [error, setError] = React.useState(null);
-  const isMounted = React.useRef(true);
+export const useFetch = url => {
+  const _timer = React.useRef(null);
+  const _controller = React.useRef(null);
+
+  const _isMounted = React.useRef(false);
+  const _isBlocked = React.useRef(false);
+
+  const [isLoading, _setIsLoading] = React.useState(false);
+  const [response, _setResponse] = React.useState(null);
+  const [error, _setError] = React.useState(null);
+
+  const isMounted = () => _isMounted.current;
+  const isUnblocked = () => !_isBlocked.current;
+
+  const setIsLoading = _isLoading => doIf(isMounted, _setIsLoading, _isLoading);
+  const setResponse = _response => doIf(isMounted, _setResponse, _response);
+  const setError = _error => doIf(isMounted, _setError, _error);
 
-  /* eslint-disable no-undef */
-  const abortController = new AbortController();
+  const reset = () => {
+    setIsLoading(false);
+    setResponse(null);
+    setError(null);
+  };
 
   React.useEffect(() => {
-    if (url) {
-      (async () => {
-        const { signal } = abortController;
-        const headers = { authorization: getAuthorizationHeader() };
-        try {
-          const response = await fetch(url, { headers, signal, ...options });
-          const { ok, status } = response;
-          if (ok) {
-            const responseData = await response.json();
-            const { error: responseError } = responseData;
-            if (responseError) throw new BugsnagError(responseError, { url, headers });
-            if (!responseData.length) throw new Error(ERROR_CODES.EMPTY_RESPONSE);
-            if (isMounted.current) setData(responseData);
-          } else {
-            switch (status) {
-              case 400:
-              default:
-                throw new Error(ERROR_CODES.CONNECTION_FAILURE);
-              case 401:
-                throw new BugsnagError(ERROR_CODES.INVALID_PASSWORD, { url, headers });
-            }
-          }
-        } catch (responseError) {
-          if (isMounted.current) {
-            if (!isValidError(responseError)) {
-              responseError.message = ERROR_CODES.CONNECTION_FAILURE;
-            }
-            setError(responseError);
-          }
-        }
-      })();
-      const timer = setTimeout(() => {
-        if (isMounted.current) setData();
-        abortController.abort();
-      }, timeout);
-      return () => {
-        isMounted.current = false;
-        clearTimeout(timer);
-        abortController.abort();
-      };
-    }
+    _isMounted.current = true;
     return () => {
-      isMounted.current = false;
+      _isMounted.current = false;
+      if (_controller.current) _controller.current.abort();
+    };
+  }, []);
+
+  const _fetch = async (path, init, opts) => {
+    const { responseHandler = res => res, errorHandler = err => err, timeout = 10000 } = opts;
```

Comment:
nit: I like having `opts = defaultOpts` here.

In `useFetch = (url, defaultOpts = DEFAULT_OPTIONS)` and having `DEFAULT_OPTIONS = { responseHandler = res => res, errorHandler = err => err, timeout = 10000 }` set outside the function?

Reply:
Also a bit clearer if these default functions are `async`?

Reply:
Both nice and sync now 😄.